new rule: HTML page language subtag matches default language (#1334)
act-rules/act-rules.github.io: _rules/html-page-lang-matches-default-ucwvc8.md

---

> id: ucwvc8

Author: There's no reason for that. I'm not sure how that practice snuck in. That was not the intent. I've been using https://www.random.org/strings/ to create my IDs for a while now.

name: HTML page language subtag matches default language
rule_type: atomic
description: |
  This rule checks that the primary language subtag of the page language matches the default language of the page
accessibility_requirements:
  wcag20:3.1.1: # Language of Page (A)
    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed
  wcag-technique:H57: # Using the language attribute on the HTML element
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
input_aspects:
  - DOM tree
  - Accessibility tree
  - CSS Styling
  - Language
acknowledgments:
  authors:
    - Wilco Fiers
---

## Applicability

This rule applies to any [document element][] if it is an `html` element for which one of the following is true:

- The [document element][] has a `lang` attribute with a value that is a [valid language tag][]; and
- The [document element][] is in a [top-level browsing context][]; and

> - The [document element][] has a [node document][] with a [content type][] of `text/html`; and

Author: I copied this from the other lang rules. We should change it there too.

- The [document element][] has a defined [default page language][].

## Expectation

For each test target, the [primary language][] of the [valid language tag][] matches the [default page language][] of the test target.

## Assumptions

> - The way to determine the default language of a page should be done by counting the number of visible words used in each language. If the default language needs to be derived in some other way, this rule may produce incorrect results.

Author: I think this is the right way to do it. It's an assumption about how to define "default language". I could do something like "The way to determine the default language of a page, as described in WCAG 2 ..." Perhaps that addresses it.

- This rule assumes that user agents and assistive technologies can programmatically determine [valid language tags](#valid-language-tag) even if these do not conform to the [BCP 47][] syntax.

- This rule assumes that [grandfathered tags][] are not used as these will not be recognized as [valid language tags](#valid-language-tag).

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [Understanding Success Criterion 3.1.1: Language of Page](https://www.w3.org/WAI/WCAG21/Understanding/language-of-page.html)
- [H57: Using language attributes on the html element](https://www.w3.org/WAI/WCAG21/Techniques/html/H57)
- [BCP 47: Tags for Identifying Languages](https://www.ietf.org/rfc/bcp/bcp47.txt)
- [The `lang` and `xml:lang` attributes](https://html.spec.whatwg.org/multipage/dom.html#the-lang-and-xml:lang-attributes)

## Test Cases

### Passed

#### Passed Example 1

This page has a [default language][default page language] of English, which matches the language indicated in the `lang` attribute.

```html

> <!DOCTYPE html>

Reviewer: I think the DOCTYPE is not needed when its html, as it will be added automatically (https://github.com/act-rules/act-rules-web/blob/3cb73b7cb5e3a39b6c342fcb7a89ab6551dcbc32/build/create-testcases.js).

Author: Wow, odd, why doesn't it check if there is a doctype before adding one in? :/ I've filed an issue. I'll take them out.

<html lang="en">
	<head>
		<title>ACT Rules Format 1.0 - Abstract</title>
	</head>
	<body>
		<p>
			The Accessibility Conformance Testing (ACT) Rules Format 1.0 defines a format for writing accessibility test
			rules. These test rules can be used for developing automated testing tools and manual testing methodologies. It
			provides a common format that allows any party involved in accessibility testing to document and share their
			testing procedures in a robust and understandable manner. This enables transparency and harmonization of testing
			methods, including methods implemented by accessibility test tools.
		</p>
	</body>
</html>
```

#### Passed Example 2

This page has a [default language][default page language] of English because most words are English. The [default page language][] matches the language indicated in the `lang` attribute.

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title>Gelukkig</title>
	</head>
	<body>

> 		<p>The Dutch word "gelukkig" has no equivalent in English.</p>

Reviewer: I am a bit worried about examples like that (due to the definition of default language).

`The word ale is from "øl"`
Both "the" (alternate spelling of the correct "te"), "ale", "is" and "from" are also Danish words (according to ordnet.dk). So this sentence has actually 5 English words and 5 Danish words, thus no default language. Although it has absolutely no meaning as a sentence in Danish.

Now, this is OK-ish, there is no default language so we don't do anything.

What I'm worried of is that we can tweak such an example enough to have a sentence with, say, 8 words that are English and 10 that are Danish, but just because some English words happen to have a completely different meaning in Danish. The sentence would be undoubtfully English, but the default language would be Danish.

Then, I guess this would be a case where the first assumption is wrong, so that is correct…

Maybe strengthen the first assumption by giving examples of other way to detect language? (letters/words frequency analysis, mutual information based distance, n-grams, …)

Author: I thought of that too. This seems fairly hypothetical though. Real world web pages don't just have a single phrase on them. I'm counting on pages having enough content on them to make things like the same word existing in multiple language not matter.

There are a number of other problems with this approach. The biggest one I have with it is that it's difficult to do accurately. If a page is half English half Danish, you can obviously word count the thing, but grabbing all the "Danish" words from the English section, is difficult to do. I explored some other ideas for this, like looking at sentences or paragraphs and counting those, but there you get into issues of how to define those. Also how do you compare two short paragraphs or two short sentences to a single long one. There's also the issue that some languages concatenate words. The same paragraph in German will have fewer words in it than the English one, just because in English words aren't concatenated.

The reason I went with this approach is because it is going to be easy to do for 99.9% of all pages. You can pretty much just eyeball it in almost every possible case, and in the cases they can't, a simple word count tool will resolve the problem. That extreme case where it is too close to call, yes this isn't ideal, but solving for that will probably take months, and is likely going to add complexity to testing this for the 99.9%. I don't think that's worth it.

This page has a [default language][default page language] of English, which is not the language indicated in the `lang` attribute.

```html
<!DOCTYPE html>
<html lang="da">
	<head>
		<title>ACT Rules Format 1.0 - Abstract</title>
	</head>
	<body>
		<p>
			The Accessibility Conformance Testing (ACT) Rules Format 1.0 defines a format for writing accessibility test
			rules. These test rules can be used for developing automated testing tools and manual testing methodologies. It
			provides a common format that allows any party involved in accessibility testing to document and share their
			testing procedures in a robust and understandable manner. This enables transparency and harmonization of testing
			methods, including methods implemented by accessibility test tools.
		</p>
	</body>
</html>
```

#### Failed Example 2

This page has a [default language][default page language] of English because most words are English. This does not match the language indicated in the `lang` attribute.

```html
<!DOCTYPE html>
<html lang="nl">
	<head>
		<title>Gelukkig</title>
	</head>
	<body>
		<p>The Dutch word "gelukkig" has no equivalent in English.</p>
	</body>
</html>
```

> #### Failed Example 3

Reviewer: Should there be a Passed Example similar to this?

This page has a [default language][default page language] of Dutch because all English text is wrapped in a `p` element with a `lang` attribute set to `en`. Dutch ("nl") is not set as the language of the page by the `lang` attribute on the `html` element.

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title>Met de kippen op stok</title>
	</head>
	<body>
		<blockquote>
			<p>"Hij ging met de kippen op stok"</p>
		</blockquote>
		<p lang="en">
			This Dutch phrase literally translates into "He went to roost with the chickens", but it means that he went to bed
			early.
		</p>
	</body>
</html>
```

> ### Inapplicable

Reviewer: Missing inapplicable examples for:
* invalid lang attribute
* not a top-level browsing context
* not a text/html content type.

Author: The other lang rules don't have this either, for a number of reasons:
- invalid lang would not satisfy the SC.
- not a top-level, I suppose we could add a page with no content, and only an iframe, but that'd give a page without a title. Not sure that's a good thing to put in here.
- not text/html; We can't guarantee these code snippets will be served with the correct MIME type. I don't think it's possible for us to do this at this time.

Reviewer: Agreed on the first two points.

> * not text/html; We can't guarantee these code snippets will be served with the correct MIME type. I don't think it's possible for us to do this at this time.

For this one, when we use ` ```xhtml ` instead of ` ```html `, we're adding a different DOCTYPE. I should be enough to serve the example with the correct type, I would say…
We do that in https://act-rules.github.io/rules/5b7ae0#inapplicable-example-4

> ### Inapplicable

Reviewer: Also, can you try and put inapplicable examples in the same order as the items in the applicability?
(not an html element, no lang, invalid lang, not a top-level browsing context, not text/html, no default language)

#### Inapplicable Example 1

This page has an undefined [default language][default page language] because it has no content or [document title][].

```html
<!DOCTYPE html>
<html></html>
```

#### Inapplicable Example 2

This is an SVG [document][document element], not an HTML document.

```svg
<svg xmlns="http://www.w3.org/2000/svg" lang="fr"></svg>
```

#### Inapplicable Example 3

This page has an undefined [default language][default page language] because it has no [document title][] all its content is wrapped in element with a `lang` attribute.

```html
<!DOCTYPE html>
<html>
	<p lang="en">
		The Accessibility Conformance Testing (ACT) Rules Format 1.0 defines a format for writing accessibility test rules.
		These test rules can be used for developing automated testing tools and manual testing methodologies. It provides a
		common format that allows any party involved in accessibility testing to document and share their testing procedures
		in a robust and understandable manner. This enables transparency and harmonization of testing methods, including
		methods implemented by accessibility test tools.
	</p>
</html>
```

#### Inapplicable Example 4

> In page has an undefined [default language][default page language] can not be defined because it can either be English or French.

Reviewer: ```suggestion
In page has an undefined [default language][default page language] because it can either be English or French.
```

---
id: ucwvc8
name: HTML page language subtag matches default language
rule_type: atomic
description: |
  This rule checks that the primary language subtag of the page language matches the default language of the page
accessibility_requirements:
  wcag20:3.1.1: # Language of Page (A)
    forConformance: true
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed
  wcag-technique:H57: # Using the language attribute on the HTML element
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: satisfied
input_aspects:
  - DOM tree
  - Accessibility tree
  - CSS Styling
  - Language
acknowledgments:
  authors:
    - Wilco Fiers
---

## Applicability

This rule applies to any [document element][] if it is an `html` element for which all of the following are true:

- The [document element][] has a `lang` attribute with a value that is a [valid language tag][]; and
- The [document element][] is in a [top-level browsing context][]; and
- The [document element][] has a [content type][] of `text/html`; and
- The [document element][] has a defined [default page language][].

## Expectation

> For each test target, the [primary language][] of the [valid language tag][] matches the [default page language][] of the test target.

Author: This is intentional. I think this reads better. Why do you think this needs to change?

> For each test target, the [primary language][] of the [valid language tag][] matches the [default page language][] of the test target.

Reviewer 2: The definition we are referring to here redirects us to the "primary language subtag" and on the other hand, primary language subtag refers to the language not the complete tag with the country. For example, primary language can be `en-GB` and `en-US` but the primary subtag would be `en`, not `GB` as it's a region where it used which specifies the language itself.
I guess I misunderstood that term here, you meant the primary language as general whether simple language subtag or language with other subtags such as `en` or `en-US` respectively.
Please correct me if I'm still misunderstanding it.

> For each test target, the [primary language][] of the [valid language tag][] matches the [default page language][] of the test target.

Author: It's a specifier. A language tag only consists of subtags, so adding it doesn't provide additional meaning. It's like saying "The child node of the node", you might as well say "the child of the node", because all children of a DOM node are themselves nodes. Both phrases mean the same thing, but the first one has unnecessary specificity in it.

If you insist, I'll take on the change, but I don't think it's necessary.

> For each test target, the [primary language][] of the [valid language tag][] matches the [default page language][] of the test target.

Reviewer 2: Well explained. No need to change it. 👍

## Assumptions

> - This rule assumes that the default human language of a page, as described in WCAG 2, can be determined by counting the number of words used in each language. If the default language needs to be derived in some other way (such as frequency analysis, mutual information based distance, …), this rule may fail while [Success Criterion 3.1.1: Language of Page](https://www.w3.org/TR/WCAG21/#language-of-page) is still satisfied.

Author: This is intentional. This isn't just true for WCAG 2.1, it is true for all of WCAG 2.

	</body>
</html>
```

#### Passed Example 3

This page has a [default language][default page language] of Dutch because all English text is wrapped in a `p` element with a `lang` attribute set to `en`. Dutch ("nl") is set as the language of the page by the `lang` attribute on the `html` element.

```html
<html lang="nl">
	<head>
		<title>Met de kippen op stok</title>
	</head>
	<body>
		<blockquote>
			<p>"Hij ging met de kippen op stok"</p>
		</blockquote>
		<p lang="en">
			This Dutch phrase literally translates into "He went to roost with the chickens", but it means that he went to bed
			early.
		</p>
	</body>
</html>
```

#### Passed Example 4

This page has a [default language][default page language] of English because the `title` element and the [accessible name][] of the `img` element are English. The [visible][] text Dutch text does not count as its parent element has a `lang` attribute.

```html
<html lang="en">
	<head>
		<title>Fireworks over Paris</title>
	</head>
	<body>
		<img src="/test-assets/shared/fireworks.jpg" alt="Fireworks over Paris" />
		<p lang="nl">
			Gelukkig nieuwjaar!
		</p>
	</body>
</html>
```

> ### Failed

Reviewer 2: I recommend having a failed example with "grandfatheredtag".

Author: Grandfather tags are part of the assumptions, we don't need an example for it. I'm also not too sure what the benefit of it would be to the rule.

	</head>
	<body>
		<blockquote>
			<p>"Hij ging met de kippen op stok"</p>
		</blockquote>
		<p lang="en">
			This Dutch phrase literally translates into "He went to roost with the chickens", but it means that he went to bed
			early.
		</p>
	</body>
</html>
```

#### Failed Example 4

This page has a [default language][default page language] of English because the `title` element and the [accessible name][] of the `img` element are English. English ("en") is not set as the language of the page by the `lang` attribute on the `html` element.

```html
<html lang="nl">
	<head>
		<title>Fireworks over Paris</title>
	</head>
	<body>
		<img src="/test-assets/shared/fireworks.jpg" alt="Fireworks over Paris" />
		<p lang="nl">
			Gelukkig nieuwjaar!
		</p>
	</body>
</html>
```

### Inapplicable

#### Inapplicable Example 1

This is an SVG [document][document element], not an HTML document.

```svg
<svg xmlns="http://www.w3.org/2000/svg" lang="fr"></svg>
```

#### Inapplicable Example 2

This page has an undefined [default language][default page language] because it has no content or [document title][].

```html
<html></html>
```

#### Inapplicable Example 3

> This page has an undefined [default language][default page language] because it has no [document title][] all its content is wrapped in element with a `lang` attribute.

Reviewer 3: ```suggestion
This page has an undefined [default language][default page language] because it has no [document title][] and all its content is wrapped in an element with a `lang` attribute.
```

```html
<html>
	<p lang="en">
		The Accessibility Conformance Testing (ACT) Rules Format 1.0 defines a format for writing accessibility test rules.
		These test rules can be used for developing automated testing tools and manual testing methodologies. It provides a
		common format that allows any party involved in accessibility testing to document and share their testing procedures
		in a robust and understandable manner. This enables transparency and harmonization of testing methods, including
		methods implemented by accessibility test tools.
	</p>
</html>
```

#### Inapplicable Example 4

This page has an undefined [default language][default page language] because it can either be English or French.

```html
<html lang="fr">
	<head>
		<title>Paul put dire comment on tape</title>
	</head>
	<body>
		<p>Paul put dire comment on tape</p>
	</body>
</html>
```

[valid language tag]: #valid-language-tag
[default page language]: #default-page-language

> [primary language]: https://tools.ietf.org/html/bcp47#section-2.2.1 'Definition of primary language subtag'

Author: As above, I think this hinders readability of the rule.

---
id: ucwvc8
name: HTML page language subtag matches default language
rule_type: atomic
description: |

>   This rule checks that the primary language subtag of the page language matches the default language of the page

Reviewer 4: Based on https://github.com/act-rules/act-rules.github.io/pull/1334/files#r464284953

Couldn't we omit "subtag" here as well?

```suggestion
  This rule checks that the primary language of the page language matches the default language of the page
```

Author: See my comment below. You do have to say subtag here. The "primary language of the page" in itself is not the same as the "primary language of the language tag"

Reviewer 4: Fair enough.

- This rule assumes that user agents and assistive technologies can programmatically determine [valid language tags](#valid-language-tag) even if these do not conform to the [BCP 47][] syntax.

- This rule assumes that [grandfathered tags][] are not used as these will not be recognized as [valid language tags](#valid-language-tag).

- This rule assumes that `iframe` title elements are not exposed to assistive technologies and so does not consider them as part of the [default page language][].

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [Understanding Success Criterion 3.1.1: Language of Page](https://www.w3.org/WAI/WCAG21/Understanding/language-of-page.html)
- [H57: Using language attributes on the html element](https://www.w3.org/WAI/WCAG21/Techniques/html/H57)
- [BCP 47: Tags for Identifying Languages](https://www.ietf.org/rfc/bcp/bcp47.txt)
- [The `lang` and `xml:lang` attributes](https://html.spec.whatwg.org/multipage/dom.html#the-lang-and-xml:lang-attributes)

## Test Cases

### Passed

#### Passed Example 1

> This page has a [default language][default page language] of English, which matches the language indicated in the `lang` attribute.

Reviewer 4: We are not explaining why this page has a default language of English. Adding "as all of its words are English" or something along these lines may help clarity and consistency.

Reviewer 2: This description is also a bit another way round with the applicability and expectation of this rule. I guess the description of examples should be according to the [test case design](https://github.com/act-rules/act-rules.github.io/blob/develop/pages/design/rule-design.md#test-cases) that is the example has or does not have or contains the necessary condition to be met and expectation. Indirectly, conditions in the case of a test rule is "Applicability" and then comes "Expectation". 
It does not mean that this description is not sufficient or incorrect but it should be according to the template we defined in [test case design](https://github.com/act-rules/act-rules.github.io/blob/develop/pages/design/rule-design.md#test-cases).
According to the description of the rule plus the applicability and expectation, this rule checks that the primary language subtag of the page language matches the default language of the page. It is similarly the same if we say the default language of this page matches the primary language subtag of the valid language tag of the page but another way round against the description of the rule.
```suggestion
This page has a [primary language][] English ("en") set in the [valid language tag][] which matches the [default page language][] that is English.
```

```html
<html lang="en">
	<head>
		<title>ACT Rules Format 1.0 - Abstract</title>
	</head>
	<body>
		<p>
			The Accessibility Conformance Testing (ACT) Rules Format 1.0 defines a format for writing accessibility test
			rules. These test rules can be used for developing automated testing tools and manual testing methodologies. It
			provides a common format that allows any party involved in accessibility testing to document and share their
			testing procedures in a robust and understandable manner. This enables transparency and harmonization of testing
			methods, including methods implemented by accessibility test tools.
		</p>
	</body>
</html>
```

#### Passed Example 2

This page has a [default language][default page language] of English because most words are English. The [default page language][] matches the language indicated in the `lang` attribute.

```html
<html lang="en">
	<head>
		<title>Gelukkig</title>
	</head>
	<body>
		<p>The Dutch word "gelukkig" has no equivalent in English.</p>
	</body>
</html>
```

#### Passed Example 3

> This page has a [default language][default page language] of Dutch because all English text is wrapped in a `p` element with a `lang` attribute set to `en`. Dutch ("nl") is set as the language of the page by the `lang` attribute on the `html` element.

Reviewer 2: Examples description varies, shouldn't it be consistent?

Author: I don't understand what you mean with this.

```html
<html lang="nl">
	<head>
		<title>Met de kippen op stok</title>
	</head>
	<body>
		<blockquote>
			<p>"Hij ging met de kippen op stok"</p>
		</blockquote>
		<p lang="en">
			This Dutch phrase literally translates into "He went to roost with the chickens", but it means that he went to bed
			early.
		</p>
	</body>
</html>
```

#### Passed Example 4

This page has a [default language][default page language] of English because the `title` element and the [accessible name][] of the `img` element are English. The [visible][] text Dutch text does not count as its parent element has a `lang` attribute.

```html
<html lang="en">
	<head>
		<title>Fireworks over Paris</title>
	</head>
	<body>
		<img src="/test-assets/shared/fireworks.jpg" alt="Fireworks over Paris" />
		<p lang="nl">
			Gelukkig nieuwjaar!
		</p>
	</body>
</html>
```

### Failed

#### Failed Example 1

> This page has a [default language][default page language] of English, which is not the language indicated in the `lang` attribute.

Reviewer 4: We are not explaining why this page has a default language of English. Adding "as all of its words are English" or something along these lines may help clarity and consistency.

Author: I don't think we need to explain that. How definitions work rarely gets explained unless it is essential for understanding the example.

Reviewer 4: I get that it is not "essential" to understand the example, but I do think it would help. Mildly thinking these explanations should be added.

#### Passed Example 2

> This page has `lang` attribute value of `en` (English), which matches the [default language of the page][default page language]. The default language is English because all but a few words are English.

Reviewer: ```suggestion
This page has a `lang` attribute value of `en` (English), which matches the [default language of the page][default page language]. The default language is English because all but a few words are English.
```

#### Passed Example 4

> This page has `lang` attribute value of `en` (English), which matches the [default language of the page][default page language]. The default language is English because the accessible texts are English, and all other text is in a `p` element with a `lang` attribute value of `nl`.

Reviewer: ```suggestion
This page has a `lang` attribute value of `en` (English), which matches the [default language of the page][default page language]. The default language is English because the accessible texts are English, and all other text is in a `p` element with a `lang` attribute value of `nl`.
```

#### Failed Example 2

> This page has `lang` attribute value of `nl` (Dutch), which does not matches the [default language of the page][default page language]. The default language is English because all but a few words are English.

Reviewer: ```suggestion
This page has a `lang` attribute value of `nl` (Dutch), which does not matches the [default language of the page][default page language]. The default language is English because all but a few words are English.
```

#### Failed Example 2

> This page has a `lang` attribute value of `nl` (Dutch), which does not matches the [default language of the page][default page language]. The default language is English because all but a few words are English.

Reviewer: ```suggestion
This page has a `lang` attribute value of `nl` (Dutch), which does not match the [default language of the page][default page language]. The default language is English because all but a few words are English.
```

#### Failed Example 4

> This page has `lang` attribute value of `nl` (Dutch), which does not match the [default language of the page][default page language]. The default language is English because the accessible texts are English, and all other text is in a `p` element with a `lang` attribute value of `nl`.

Reviewer: ```suggestion
This page has a `lang` attribute value of `nl` (Dutch), which does not match the [default language of the page][default page language]. The default language is English because the accessible texts are English, and all other text is in a `p` element with a `lang` attribute value of `nl`.
```

</html>
```

#### Failed Example 5

> This page has `lang` attribute value of `nl` (Dutch), which does not match the [default language of the page][default page language]. The default language is English because the accessible name of the `img` element is English. The `lang` attribute on the `p` element is effectively ignored.

Reviewer: ```suggestion
This page has a `lang` attribute value of `nl` (Dutch), which does not match the [default language of the page][default page language]. The default language is English because the accessible name of the `img` element is English. The `lang` attribute on the `p` element is effectively ignored.
```
